HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,192 @@
+            break
+
+    return retP
+
+# Input is graphRDD
+# Returns (numComponents, maxCompSize)
+# Just reusing the functions above, but would be more efficient to not store the path
+def findComponents(graphRDD):
+    maxCompSize = -1
+    numComponents = 0
+    while True:
+        numComponents += 1
+        firstE = graphRDD.take(1)
+        # Reference: http://stackoverflow.com/questions/28454357/spark-efficient-way-to-test-if-an-rdd-is-empty
+        if not firstE:
+            break
+        startN = firstE[0][0]
+        graphRDD = graphRDD.map(lambda (k,v): (k,v) if (k!=startN) else (k, (0, v[1], 1, [k])), preservesPartitioning=True)
+        while True: # While gray nodes still exist
+            updates.value = 0
+            graphRDD = graphRDD.flatMap(processNode)\
+                            .reduceByKey(combineNode)
+            graphRDD.foreach(lambda (k,v): updates.add(int(v[2]==1)))
+            if updates.value == 0:
+                break
+        # Count the seen nodes
+        curCSize = graphRDD.filter(lambda (k,v): (v[0]!=INF))\
+                            .count()
+        maxCompSize = max(curCSize, maxCompSize)
+        graphRDD = graphRDD.filter(lambda (k,v): (v[0]==INF)) # Select unseen nodes
+    return (numComponents, maxCompSize)
+
+# node_index | current_dist | adj_list | color | path
+def setupNode(s):
+    s = s.split()
+    return ( int(s[0][:-1]), (INF, [int(x) for x in s[1:]], 0, []) )
+
+def emitN(s):
+    ret = []
+    s = s.split()
+    cur = int(s[0][:-1])
+    nbs = [int(x) for x in s[1:]]
+    for nb in nbs:
+        ret.append((nb, [cur]))
+    ret.append( (cur, nbs) )
+    return ret
+
+def onlyBothDir(v):
+    ret = []
+    count = {}
+    for c in v:
+        if c not in count:
+            count[c] = 1
+        else:
+            ret.append(c)
+    return (k, ret)
+
+def oneDir(v):
+    ret = []
+    seen = {}
+    for c in v:
+        if c not in seen:
+            ret.append(c)
+            seen[c] = True
+    return (k, ret)
+
+def getIdx(RDD, s):
+    ret = RDD.filter(lambda (k,v): (k==s)).take(1)
+    if not ret:
+        return -1
+    else:
+        return ret[0][1] + 1
+
+def resolvePath(RDD, p):
+    return [RDD.filter(lambda (k,v): v == i-1).take(1)[0][0] for i in p]
+
+links = sc.textFile('s3://Harvard-CS205/wikipedia/links-simple-sorted.txt')
+links = links.map(setupNode)
+
+page_names = sc.textFile('s3://Harvard-CS205/wikipedia/titles-sorted.txt')
+page_names = page_names.zipWithIndex().partitionBy(numPart).cache()
+
+# Get the indices for the search
+kb_idx = getIdx( page_names, "Kevin_Bacon" )
+hu_idx = getIdx( page_names, "Harvard_University" )
+
+# Get the paths for our search
+kb2hu = getPath(links, kb_idx, hu_idx)
+ret1 = resolvePath(page_names, kb2hu)
+print 'Kevin Bacon => Harvard University'
+print ret1
+
+hu2kb = getPath(links, hu_idx, kb_idx)
+ret2 = resolvePath( page_names, hu2kb )
+print 'Harvard University => Kevin Bacon'
+print ret2
+
+
```

Comment:
Thanks for taking the time to include comments - it makes your code very readable. :+1: